src/herder/TxSetFrame.h

```diff
@@ -187,6 +189,157 @@ class TxSetXDRFrame : public NonMovableOrCopyable
+//   'stages' (`TxStageFrame`). A stage has to be executed after every
+//   transaction in the previous stage has been applied.
+// - A 'stage' (`TxStageFrame`) consists of several parallel 'threads'
+//   (`TxThreadFrame`). Transactions in different 'threads' are independent of
+//   each other and can be applied in parallel.
+// - A 'thread' (`TxThreadFrame`) consists of transactions that should
+//   generally be applied sequentially. However, not all the transactions in
+//   the thread are necessarily conflicting with each other; it is possible
+//   that some, or even all transactions in the thread structure can be applied
+//   in parallel with each other (depending on their footprints).
+//
+// This structure mimics the XDR structure of the `ParallelTxsComponent`.
+using TxThreadFrame = TxFrameList;
+using TxStageFrame = std::vector<TxThreadFrame>;
+using TxStageFrameList = std::vector<TxStageFrame>;
+
+// Alias for the map from transaction to its inclusion fee as defined by the
+// transaction set.
+using InclusionFeeMap =
+    std::unordered_map<TransactionFrameBaseConstPtr, std::optional<int64_t>>;
+
+// `TxSetPhaseFrame` represents a single phase of the `ApplicableTxSetFrame`.
+//
+// Phases can only be created as a part of the `ApplicableTxSetFrame` and thus
+// don't have any public constructors.
+//
+// Phase may either wrap the corresponding `TransactionPhase` XDR for
+// generalized transactions sets, or represent all the transactions in the
+// 'legacy' transaction set (which is considered to have only a single phase).
+//
+// This does not assume any specific order of transactions by default - the
+// phase in 'apply' order has to be explicitly requested from the parent
+// `ApplicableTxSetFrame` via `getPhasesInApplyOrder` method.
+class TxSetPhaseFrame
+{
+  public:
+    // Returns true when this phase can be applied in parallel.
+    // Currently only Soroban phase can be parallel, and only starting from
+    // PARALLEL_SOROBAN_PHASE_PROTOCOL_VERSION protocol
+    bool isParallel() const;
+
+    // Returns the parallel stages of this phase.
+    //
+    // This may only be called when `isParallel()` is true.
+    TxStageFrameList const& getParallelStages() const;
+    // Returns all the transactions in this phase if it's not parallel.
+    //
+    // This may only be called when `isParallel()` is false.
+    TxFrameList const& getNonParallelTxs() const;
+
+    // Serializes this phase to the provided XDR.
+    void toXDR(TransactionPhase& xdrPhase) const;
+
+    // Iterator over all transactions in this phase.
+    // The order of iteration is defined by the parent `ApplicableTxSetFrame`.
+    // If the phase is sorted for apply, then the iteration order can be used
+    // to determine a stable index of every transaction in the phase, even if
+    // the phase is parallel and can have certain transaction applied in
+    // arbitrary order.
+    class Iterator
```

Comment:
This changeset in general has to precede the actual application changes, because in order to apply a tx set one needs to have an appropriate schedule. Moreover, the purpose of iterator is to actually hide the internal structure of the phase for the scenarios where we don't care about the parallel execution at all (e.g. when we just want to check if every tx in the phase is valid etc.).